concierge_scheduler/concierge_zabbix.py

```diff
@@ -307,32 +380,96 @@ def import_actions(self):
+    def compare_ids(self, component):
+        """
+        Compare the ids and hashes of components for imported configuration.
+        Remove any components that are in the destination server but not in the imported configuration
+        :param component: Which component to compare the data of
         """
-        import backed up media
 
-        :param component: what media to import
-        :return:
+        component_id = self.id_mapping[component]['id']
+        ids_to_delete = []
+        for item in self.original_ids[component]:
+            import_id = item[component_id]
+            if import_id in self.dest_ids[component]:
+                if item["hash"] != self.dest_ids[component][import_id][1]:
+                    ids_to_delete.append(import_id)
+
+        for del_id in list(ids_to_delete):
+            print("Deleting '{}' ids: {}".format(component, del_id))
+            getattr(self.zbx_client, self.id_mapping[component]['api_name']).delete(del_id)
+
+    def import_components(self, component):
+        """
+        Method for importing components which cannot be configured using a configuration import.
+        If the component exists in destination server it is updated
+        :param component: Component to import
         """
-        # ToDo: done because default DB install creates some mediatypes.
-        # We could run self.zbx_client.mediatype.update() instead, then handle
-        # handle the exception and only run create() if it fails
-        # 
-        self.zbx_client.mediatype.delete("1", "2", "3")
         import_file = '{}/{}.json'.format(self.data_dir, component)
-        with open(import_file, 'r') as f:
-            component_data = f.read()
-            mediatypes = json.loads(component_data)
-            for mediatype in mediatypes:
+        component_method = self.id_mapping[component]['api_name']
+        file = open(import_file).read()
+        if json.loads(file):
+            for item in json.loads(file):
+                if component in self.keys_to_remove:
+                    self.__remove_keys(item, self.keys_to_remove[component])
                 try:
-                    self.zbx_client.mediatype.update(mediatype)
+                    getattr(self.zbx_client, component_method).create(item)
                 except ZabbixAPIException:
-                    self.zbx_client.mediatype.create(mediatype)
+                    getattr(self.zbx_client, component_method).update()
 
-    def get_new_ids(self):
+    def get_id_maps(self):
+        """
+        Query Zabbix server to get template, hostgroup, host and mediatype objects.
+        These are added to map, with structure {component: {id: (name/host, hash), id: (...), ...}, component: ... }
+        :return: map of components.
+        """
+        data = defaultdict(dict)
+        # data = {'templates': {}, 'hostgroups': {}, 'hosts': {}, 'mediatypes': {} }
+        for template in self.zbx_client.template.get(output="extend"):
```

Comment:
seems like quite a bit od duplication here. Within this method but also in the `get_simple_ids` one. Worth creating a method for all the duplication?

Reply:
changed `get_id_maps` to iterate through specified list of components, so have also removed `get_new_maps()` function.

`get_simple_ids` makes a dictionary of lists that gets written to a file, this function is used for `backup_config`.

`get_id_maps` makes a dictionary of dictionaries stored only as a variable, that is used in `restore_config` methods.

So have kept both of these method but updated them to remove repetition for the components